### PR TITLE
Implement sponge block

### PR DESCRIPTION
**Pull request by lhochbaum**
Are ya ready, kids?

### CODE REVIEW

**Comment by TwistedAsylumMC** [comment]
Could you add the functionality of dry sponges?

File: dragonfly/block/sponge.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	if s.Wet {
+		return "minecraft:sponge", map[string]interface{}{"sponge_type": "wet"}
+	}
+	return "minecraft:sponge", map[string]interface{}{"sponge_type": "dry"}
+}
+
+// UseOnBlock ...
+func (s Sponge) UseOnBlock(pos world.BlockPos, face world.Face, v mgl64.Vec3, w *world.World, user item.User, ctx *item.UseContext) (used bool) {
+	// the sponge is already wet so it cannot consume any more water.
+	if !s.Wet {
```
Comment (Sandertv) [inline dragonfly/block/sponge.go:44]: On bedrock edition it seems to work a little different. Sponges only absorb water if they are directly adjacent to a water block. They can also absorb water if water flows adjacent to them after having been placed and if they're still dry. You can use the NeighbourUpdateTick method for that.

File: dragonfly/block/sponge.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+			if len(queue) == 0 {
+				break
+			}
+
+			// pop the next tuple entry from the queue.
+			next := queue[0]
+			queue = queue[1:]
+
+			next.block.Neighbours(func(neighbour world.BlockPos) {
+				block := w.Block(neighbour)
```
Comment (Sandertv) [inline dragonfly/block/sponge.go:66]: Use `w.Liquid(neighbour)` to account for waterlogged blocks here.

File: dragonfly/block/sponge.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+				if _, isWater := block.(Water); isWater {
+					w.SetBlock(neighbour, Air{})
```
Comment (Sandertv) [inline dragonfly/block/sponge.go:68]: Use `w.SetLiquid(neighbour, nil)` to account for waterlogged blocks here.

File: dragonfly/block/sponge.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	}
+	return "minecraft:sponge", map[string]interface{}{"sponge_type": "dry"}
+}
+
+// UseOnBlock ...
+func (s Sponge) UseOnBlock(pos world.BlockPos, face world.Face, v mgl64.Vec3, w *world.World, user item.User, ctx *item.UseContext) (used bool) {
+	// the sponge is already wet so it cannot consume any more water.
+	if !s.Wet {
+		// tuple binds a world.BlockPos to its distance from the sponge's position.
+		type tuple struct {
```
Comment (Sandertv) [inline dragonfly/block/sponge.go:46]: Consider naming this a bit different, like `distanceToSponge`.

File: dragonfly/block/sponge.go (hunk 1 above)
Comment (Sandertv) [reply dragonfly/block/sponge.go:44]: You might want to pull the absorbing behaviour out of this method into a new method and call it on NeighbourUpdateTick and on placement, while checking if there is water directly adjacent to it.

File: dragonfly/block/sponge.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+		// at least one water block has been consumed, so the sponge is no longer dry.
+		if replaced > 0 {
+			s.Wet = true
+		}
+	}
+
+	pos, face, used = firstReplaceable(w, pos, face, s)
+	if !used {
+		return
+	}
```
Comment (TwistedAsylumMC) [inline dragonfly/block/sponge.go:90]: Might be worth putting this before removing the water

File: dragonfly/block/sponge.go [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+// absorbed.
+func (s Sponge) UseOnBlock(pos world.BlockPos, face world.Face, _ mgl64.Vec3, w *world.World, user item.User, ctx *item.UseContext) (used bool) {
+	pos, face, used = firstReplaceable(w, pos, face, s)
+	if !used {
+		return
+	}
+
+	place(w, pos, s, user, ctx)
+
+	// the sponge is dry, so it can absorb nearby water.
```
Comment (Sandertv) [inline dragonfly/block/sponge.go:51]: Please capitalise this sentence.

File: dragonfly/block/sponge.go [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+		}
+	}
+
+	return placed(ctx)
+}
+
+// NeighbourUpdateTick checks for nearby water flow. If water could be found and the sponge is dry, it will absorb the
+// water and be flagged as wet.
+func (s Sponge) NeighbourUpdateTick(pos, _ world.BlockPos, w *world.World) {
+	// the sponge is dry, so it can absorb nearby water.
```
Comment (Sandertv) [inline dragonfly/block/sponge.go:66]: Please capitalise this sentence.

File: dragonfly/block/sponge.go [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+	if !s.Wet {
+		if s.absorbWater(pos, w) > 0 {
+			// water has been absorbed, so we flag the sponge as wet.
```
Comment (Sandertv) [inline dragonfly/block/sponge.go:69]: Please capitalise this sentence.

File: dragonfly/block/sponge.go [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+	// distanceToSponge binds a world.BlockPos to its distance from the sponge's position.
+	type distanceToSponge struct {
+		block    world.BlockPos
+		distance int32
+	}
+
+	queue := make([]distanceToSponge, 0)
+	queue = append(queue, distanceToSponge{pos, 0})
+
+	// a sponge can only absorb up to 65 water blocks.
```
Comment (Sandertv) [inline dragonfly/block/sponge.go:89]: Please capitalise this sentence.

File: dragonfly/block/sponge.go [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+	replaced := 0
+	for replaced < 65 {
+		if len(queue) == 0 {
+			break
+		}
+
+		// pop the next distanceToSponge entry from the queue.
```
Comment (Sandertv) [inline dragonfly/block/sponge.go:96]: Please capitalise this sentence.

File: dragonfly/block/sponge.go [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+					w.SetLiquid(neighbour, nil)
+					replaced++
+					if next.distance < 7 {
+						queue = append(queue, distanceToSponge{neighbour, next.distance + 1})
+					}
+				}
+			} else if _, isAir := w.Block(neighbour).(Air); isAir {
+				if next.distance < 7 {
+					queue = append(queue, distanceToSponge{neighbour, next.distance + 1})
+				}
```
Comment (Sandertv) [inline dragonfly/block/sponge.go:114]: Not sure if this is intended. Following the sponge behaviour in-game, absorbing only happens for water that is connected to other water, so if air is encountered, this should probably stop.

**Comment by Sandertv** [comment]
It also seems like in vanilla, a BreakBlock particle at the position of the sponge + 0.5 on all axes is sent with the block being water when a dry sponge turns into a wet sponge. Let me know on Discord if you need help with that.

**Pull request merged**